### PR TITLE
Mn 2017 07 remove multitable inheritance

### PR DESCRIPTION
**Pull request by MagdaN**
In this PR I tried to remove multitable inheritance from ideas. As I just wanted to test, I did not create migrations but if we decide to go this way we could still create them. So far, I only adapted the tests to see if this is working. I used Postgres.

Findings so far:

- I can delete a proposal (which keeps the Idea and the archived Idea)

- I can not delete an Idea for which a proposal has beed created before ()

### CODE REVIEW

**Comment by MagdaN** [comment]
@slomo I removed the inheritance from Item in IdeaSketchArchive, which solves the problem, yeah! But now, if I delete a proposal, everything else is kept (Idea, IdeaSketchArchive). If I delete the Idea, the proposal also gets deleted but of course the IdeaSketchArchive is kept. Final Fix could be to move the oneToOne Field between Proposal to IdeaSketchArchive from Proposal to IdeaSketchArchive.

File: tests/ideas/test_filters.py [hunk 1]
```diff
@@ -4,10 +4,15 @@
...
-    idea_sketch_factory()
-    idea_sketch_factory()
-    proposal_factory()
+def test_idea_list_view(idea_sketch_factory, idea_factory, proposal_factory):
+
+    idea1 = idea_factory(visit_camp=False)
+    idea2 = idea_factory(visit_camp=False)
+    idea3 = idea_factory(visit_camp=True)
+
+    idea_sketch_factory(idea=idea1)
```
Comment (slomo) [inline tests/ideas/test_filters.py:13]: This could probably be `idea_sketch_factory(idea__visit_camp=False)`

File: apps/ideas/models/__init__.py [hunk 2]
```diff
@@ -74,23 +75,63 @@ def type(self):
             return IdeaSketch._meta.verbose_name.title()
 
 
-class IdeaSketch(Idea, AbstractCollaborationCampSection):
+class IdeaSketch(AbstractCollaborationCampSection):
+
+    idea = models.OneToOneField(Idea)
+
+    @property
```
Comment (slomo) [inline apps/ideas/models/__init__.py:82]: There are a lot of attributes from idea that we want to have on the sketch or proposal, so maybe we should dare to something more radical and overwrite `__getattr__` (see [python docs](https://docs.python.org/3/reference/datamodel.html#object.__getattr__))

```
def __getattr__(self, name):
    if hasattr(self.idea, name):
        return getattr(self.idea, name)
   else:
       return super().__getattr(name)
```

In theory we could also do it for `__setattr__` and than automatically call `self.idea.save` from the `save` method, but I think that might be to radical.

File: apps/ideas/models/__init__.py [hunk 3]
```diff
@@ -74,23 +75,63 @@ def type(self):
...
+    def creator(self):
+        return self.idea.creator
+
+    @property
+    def collaborators(self):
+        return self.idea.collaborators
+
+    @property
+    def type(self):
```
Comment (slomo) [inline apps/ideas/models/__init__.py:91]: This becomes way nicer now I like it. We could also use `self._meta` here and than even put into some common base class.

**Review by slomo — CHANGES_REQUESTED** [review]
I would say that was a success. The code looks a lot cleaner we rely less on Django features that we don't understand or are in transparent to us. What is your opinion?

I am wondering wether the IdeaSketchArchive should not only backup the base idea and reference idea sketch model for the additional fields.

Here are some open issue that I see:

- submit notification for idea sketches is empty
- edit for idea sketches crashes
- submit notification for proposal is empty
- edit for proposal crashes
- I suspect if you update basic fields of a proposal, that will not create an update action for the proposal (not important anyhow)

Those are a lot of big issues that were not intended to solve with this PR, but we need to solve before bringing this live.

I would propose the following:
 - we apply the ideaSketchArchive factory to our current master
 - next we decide finally if we want to go this step
 - afterwards we merge the basic changes from this PR
 - finally we fix the list of issues above

**Comment by MagdaN** [comment]
@slomo we should take a look at our tests - this PR still breaks a lot of stuff without making the tests fail.

But to sum this up, I think, we agreed on using the PR with the updated django version and add the fixed IdeaSketchArchived Model from this PR.

**Comment by slomo** [comment]
Close for now, might be revisited later

**Pull request closed without merging**